// atm-common/src/defaults.js
module.exports = {
    plugins: {
        parser: {
            coffee: {
                name: 'coffee-script',
                options: {}
            },
            ts: {
                name: 'typescript',
                options: {}
            },
            sass: {
                name: 'node-sass',
                options: {}
            },
            less: {
                name: 'less-2.x',
                options: {}
            },
            stylus: {
                name: 'stylus',
                options: {}
            },
            transform: {
                name: 'transform-id',
                options: {}
            }
        },
        optimizer: {
            js: {
                name: 'uglify-js',
                options: {}

            },
            css: {
                name: 'clean-css',
                options: {}
            },
            png: {
                name: 'png-compressor',
                options: {}
            }
        },
        postprocessor: {
            ap: {
                name: 'autoprefixer',
                options: {}
            }
        }
    },
    builds: [

		// dist目录下不产出
		{
			glob: '/dist/**',
			release: false
		},
        {
            glob: '**.js',
            plugin: {
                parser: ['transform'],
                optimizer: ['js']
            }
        },
        {
            glob: '**.{es,es6,jsx,ts,tsx}',
            plugin: {
                parser: ['typescript', 'transform'],
                optimizer: ['js']
            },
            rExt: '.js'
        },
        {
            glob: '**.coffee',
            plugin: {
                parser: ['coffee', 'rid'],
                optimizer: ['js']
            },
            rExt: '.js'
        },
        {
            glob: '**.{ts,tsx}',
            plugin: {
                parser: ['ts', 'rid'],
                optimizer: 'js'
            },
            rExt: '.js'
        },
        {
            glob: '**.css',
            plugin: {
                parser: 'rid',
                optimizer: 'css',
                postprocessor: 'ap'
            }
        },
        {
            glob: '**.{sass,scss}',
            plugin: {
                parser: ['sass', 'rid'],
                optimizer: 'css',
                postprocessor: 'ap'
            },
            rExt: '.css'
        },
        {
            glob: '**.less',
            plugin: {
                parser: ['less', 'rid'],
                optimizer: 'css',
                postprocessor: 'ap'
            },
            rExt: '.css'
        },
        {
            glob: '**.styl',
            plugin: {
                parser: ['style', 'rid'],
                optimizer: 'css',
                postprocessor: 'ap'
            }
        }
    ]
}
