// atm-site/src/defaults.js
module.exports = {
    configs: {
        stable: true,
        debug: {
            param: 'debugId',
            importCss: false
        },
        loaderId: 'parser:parser',
        alias: {},
        deps: {},
	    plugins: {
	        parser: {
	            coffee: {
	                name: 'coffee-script',
					allowClose: false,
	                options: {},
					devOptions: {}
	            },
	            typescript: {
	                name: 'typescript',
					allowClose: false,
	                options: {},
					devOptions: {}
	            },
	            sass: {
	                name: 'node-sass',
					allowClose: false,
	                options: {},
					devOptions: {}
	            },
	            less: {
	                name: 'less-2.x',
					allowClose: false,
	                options: {},
					devOptions: {}
	            },
	            stylus: {
	                name: 'stylus',
					allowClose: false,
	                options: {},
					devOptions: {}
	            },
				transform: {
					name: 'transform',
					allowClose: false,
					options: {},
					devOptions: {}
				}
	        },
	        optimizer: {
	            js: {
	                name: 'uglify-js',
					allowClose: true,
	                options:{
						sourceMap: true
					}
	            },
	            css: {
	                name: 'clean-css',
					allowClose: true,
	                options: {}
	            },
	            png: {
	                name: 'png-compressor',
					allowClose: true,
	                options: {}
	            }
	        },
	        postprocessor: {
	            autoprefixer: {
	                name: 'autoprefixer',
					allowClose: false,
	                options: {}
	            }
	        }
	    },
	    builds: [
			// 配置所有文件的默认构建规则
			{
				glob: '**',
				parser: ['transform']
			},
			// atmjs/config.json目录不产出
			{
				glob: '/atmjs/**',
				release: false
			},
	        {
	            glob: '/(**).js',
	            plugins: {
	                parser: ['transform'],
	                optimizer: ['js']
	            },
				id: '$1',
				isMod: true
	        },
	        {
	            glob: '/(**).{es,es6,jsx,ts,tsx}',
	            plugins: {
	                parser: ['typescript', 'transform'],
	                optimizer: ['js']
	            },
				id: '$1',
				isMod: true,
	            rExt: '.js'
	        },
	        {
	            glob: '/(**).coffee',
	            plugins: {
	                parser: ['coffee', 'transform'],
	                optimizer: ['js']
	            },
				id: '$1',
				isMod: true,
	            rExt: '.js'
	        },
	        {
	            glob: '/**.css',
	            plugins: {
					parser: ['transform'],
	                optimizer: ['css'],
	                postprocessor: ['autoprefixer']
	            }
	        },
	        {
	            glob: '/(**).{sass,scss}',
	            plugins: {
	                parser: ['sass', 'transform'],
	                optimizer: ['css'],
	                postprocessor: ['autoprefixer']
	            },
				id: '$1.css',
	            rExt: '.css'
	        },
	        {
	            glob: '/(**).less',
	            plugins: {
	                parser: ['less', 'transform'],
	                optimizer: ['css'],
	                postprocessor: ['autoprefixer']
	            },
				id: '$1.css',
	            rExt: '.css'
	        },
	        {
	            glob: '/(**).styl',
	            plugins: {
	                parser: ['style', 'transform'],
					optimizer: ['css'],
	                postprocessor: ['autoprefixer']
	            },
				id: '$1.css'
	        },
			{
				glob: '**.png',
				plugins: {
					optimizer: ['png']
				}
			}
	    ]
    },
    settings: {
		charset: 'utf-8',
		name: '',				// 站点名称
        libName: 'lib',			// 开源文件夹名称
        together: true,
        jsonpName: 'callback',
        timeoutName: 'timeout',
        styleTimeout: 0,
		targets: {
			dev: {
				hash: false,
				closePlugins: {
					optimizer: ['js', 'css', 'png']
				},
				domain: 'http://**/*',
	            // 清单文件的根路径,可用占位符{{site}}(开发目录,即站点目录)和{{target}}(构建目标)
	            //manifests: '{{site}}/../manifests/{{target}}',
				assetsRoot: '{{sitedir}}',
				assetsOffset: 'atmdist',
				manifestRoot: '{{sitedir}}',
				manifestsOffset: 'atmdist/manifest-{{version}}.json',
				mapsRoot: '{{sitedir}}',
				mapsOffset: 'atmdist/map-{{version}}.json'
			}
		},
        operates: {
            build: {
                status: true,
                name: '构建',
                color: 'green'
            },
            map: {
                status: true,
                name: '地图',
                color: 'blue'
            },
            deploy: {
                status: true,
                name: '部署',
                color: 'brown'
            },
            publish: {
                status: true,
                name: '上线',
                color: 'orange'
            },
            build_map: {
                status: true,
                name: '构建+地图',
                color: 'blue'
            },
            build_map_deploy: {
                status: true,
                name: '构建+地图+部署',
                color: 'brown'
            },
            build_map_deploy_publish: {
                status: true,
                name: '构建+地图+部署+上线',
                color: 'orange'
            },
            map_deploy: {
                status: false,
                name: '地图+部署',
                color: 'brown'
            },
            map_deploy_publish: {
                status: false,
                name: '地图+部署+上线',
                color: 'orange'
            },
            deploy_publish: {
                status: false,
                name: '部署+上线',
                color: 'orange'
            }
        }
    }
}
