// atm-site/src/defaults.js
module.exports = {
    configs: {
        stable: true,
        debug: {
            param: 'debugId',
            importCss: false
        },
        loaderId: 'parser:parser',
        alias: {},
        deps: {},
        plugin: {
            babel: {
                blacklist: ['regenerator'],
                stage: 3
            },
            coffee: {},
            ts: {},
            sass: {},
            less: {},
            stylus: {},
            uglifyjs: {},
            cleancss: {},
            pngcompressor: {},
            autoprefixer: {}
        }
    },
    settings: {
        libName: 'lib',
        together: true,
        jsonpName: 'callback',
        timeoutName: 'timeout',
        styleTimeout: 0,
        operates: {
            build: {
                status: true,
                name: '构建',
                color: 'green'
            },
            map: {
                status: true,
                name: '地图',
                color: 'blue'
            },
            deploy: {
                status: true,
                name: '部署',
                color: 'brown'
            },
            publish: {
                status: true,
                name: '上线',
                color: 'orange'
            },
            build_map: {
                status: true,
                name: '构建+地图',
                color: 'blue'
            },
            build_map_deploy: {
                status: true,
                name: '构建+地图+部署',
                color: 'brown'
            },
            build_map_deploy_publish: {
                status: true,
                name: '构建+地图+部署+上线',
                color: 'orange'
            },
            map_deploy: {
                status: false,
                name: '地图+部署',
                color: 'brown'
            },
            map_deploy_publish: {
                status: false,
                name: '地图+部署+上线',
                color: 'orange'
            },
            deploy_publish: {
                status: false,
                name: '部署+上线',
                color: 'orange'
            }
        }
    }
}
